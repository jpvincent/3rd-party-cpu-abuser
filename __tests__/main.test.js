const cpuAbuser = require('../index')
const mockPath = '__tests__/__mocks__/trace-vanity-mobile.json'
const mockResultDomains = require('./__mocks__/output-vanity-mobile-domains.json')
const mockResultUrl = require('./__mocks__/output-vanity-mobile-url.json')

jasmine.DEFAULT_TIMEOUT_INTERVAL = 15000

describe('Common behaviour', () => {
  it('should throw an error if file does not exist', () => {
    expect(() => cpuAbuser.data({
      file: 'non-existent'
    })).toThrow()
  })
})


let resultDefault

describe('Node API with JSON results : default options', () => {
  let result = cpuAbuser.data({
    file: mockPath
  })
  resultDefault = result // for next test to compare

  it('should have a JSON structure', () => {
    //console.log(result)
    expect( result.totalCPUTime ).toBeDefined()
    expect( result.cpuTimePerDomain ).toBeDefined()
  })

  it('timings should be Number', () => {
    //console.log(result)
    expect(result.totalCPUTime).toEqual(expect.any(Number))
    expect(result.cpuTimePerDomain['pagead2.googlesyndication.com']).toEqual(expect.any(Number))
  })

  it('should calculate totalCPUTime', () => {
    // console.log(result)
    expect(result.totalCPUTime).toBe(mockResultDomains.totalCPUTime)
  })

  it('should calculate CPU time on a domain basis', () => {
    // console.log(result)
    expect(result.cpuTimePerDomain['pagead2.googlesyndication.com']).toBeDefined()
    expect(result.cpuTimePerDomain['pagead2.googlesyndication.com']).toBe(mockResultDomains.cpuTimePerDomain['pagead2.googlesyndication.com'])
  })

  it('trace-vanity-mobile.json analysis should exactly match the saved results', () => {
    //console.log(result)
    expect(result).toEqual(mockResultDomains)
    expect(result).toMatchSnapshot()
  })

})

describe('results as console table : default options', () => {
  let result = cpuAbuser.toTableConsole({
    file: mockPath
  })

  it('should match the previous mock', () => {
    expect(result).toMatchSnapshot()
  })

})

describe('Node API with JSON results : minTime=10', () => {
  let result = cpuAbuser.data({
    file: mockPath,
    minTime: 10
  })

  it('Changing minTime should return more results', () => {
    expect(result.totalOffendersDomains).toBeGreaterThan(resultDefault.totalOffendersDomains)
  })

})

describe('Node API with JSON results : minTime=0', () => {
  let result = cpuAbuser.data({
    file: mockPath,
    minTime: 0
  })

  it('Changing minTime should return more results', () => {
    expect(result.totalOffendersDomains).toBeGreaterThan(resultDefault.totalOffendersDomains)
  })

})

describe('Node API : groupBy=URL', () => {
  let result = cpuAbuser.data({
    file: mockPath,
    groupBy: 'URL'
  })

  it('should calculate CPU time on a per URL basis', () => {
    // console.log(result)
    expect(result.cpuTimePerDomain['https://pagead2.googlesyndication.com/pagead/osd.js']).toBeDefined()
    expect(result.cpuTimePerDomain['https://pagead2.googlesyndication.com/pagead/osd.js']).toBe(mockResultUrl.cpuTimePerDomain['https://pagead2.googlesyndication.com/pagead/osd.js'])
  })

  it('should calculate totalCPUTime', () => {
    // console.log(result)
    expect(result.totalCPUTime).toBe(mockResultUrl.totalCPUTime)
  })


})
