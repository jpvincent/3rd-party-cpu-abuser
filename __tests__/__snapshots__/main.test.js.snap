// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Node API with JSON results : default options trace-vanity-mobile.json analysis should exactly match the saved results 1`] = `
Object {
  "cpuTimePerDomain": Object {
    "adyoulike.omnitagjs.com": 347.20399993658066,
    "cdns.eu1.gigya.com": 224.9640000462532,
    "cdns.gigya.com": 601.3290000259876,
    "connect.facebook.net": 411.74199998378754,
    "fw.adsafeprotected.com": 560.256000071764,
    "m.vanityfair.fr": 1062.517999947071,
    "pagead2.googlesyndication.com": 169.76800015568733,
    "pcotag.azureedge.net": 396.0799999833107,
    "platform.twitter.com": 267.3179999291897,
    "secure-ds.serving-sys.com": 207.50500014424324,
    "securepubads.g.doubleclick.net": 801.206000238657,
    "widgets.outbrain.com": 427.435999751091,
    "www.googletagmanager.com": 192.87200009822845,
    "z.moatads.com": 437.3180001080036,
  },
  "totalCPUTime": 7189.355000168085,
  "totalDomains": 55,
  "totalOffendersDomains": 14,
}
`;

exports[`results as console table : default options should match the previous mock 1`] = `
"[90m┌──────────────────────────┬─────────┐[39m
[90m│[39m[31m Total CPU busy time (ms) [39m[90m│[39m 7189.36 [90m│[39m
[90m├──────────────────────────┼─────────┤[39m
[90m│[39m[31m Total number of domains  [39m[90m│[39m 55      [90m│[39m
[90m├──────────────────────────┼─────────┤[39m
[90m│[39m[31m Number of big offenders  [39m[90m│[39m 14      [90m│[39m
[90m└──────────────────────────┴─────────┘[39m
[90m┌───────────────┬────────────────────────────────┐[39m
[90m│[39m[31m CPU Time (ms) [39m[90m│[39m[31m domain name                    [39m[90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 1062.52       [90m│[39m m.vanityfair.fr                [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 801.21        [90m│[39m securepubads.g.doubleclick.net [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 601.33        [90m│[39m cdns.gigya.com                 [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 560.26        [90m│[39m fw.adsafeprotected.com         [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 437.32        [90m│[39m z.moatads.com                  [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 427.44        [90m│[39m widgets.outbrain.com           [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 411.74        [90m│[39m connect.facebook.net           [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 396.08        [90m│[39m pcotag.azureedge.net           [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 347.20        [90m│[39m adyoulike.omnitagjs.com        [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 267.32        [90m│[39m platform.twitter.com           [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 224.96        [90m│[39m cdns.eu1.gigya.com             [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 207.51        [90m│[39m secure-ds.serving-sys.com      [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 192.87        [90m│[39m www.googletagmanager.com       [90m│[39m
[90m├───────────────┼────────────────────────────────┤[39m
[90m│[39m 169.77        [90m│[39m pagead2.googlesyndication.com  [90m│[39m
[90m└───────────────┴────────────────────────────────┘[39m"
`;
